core/tee/fs_dirfile.c

```diff
@@ -238,6 +238,36 @@ TEE_Result tee_fs_dirfile_find(struct tee_fs_dirfile_dirh *dirh,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result tee_fs_dirfile_new(struct tee_fs_dirfile_dirh *dirh,
+				     struct tee_fs_dirfile_fileh *dfh)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct dirfile_entry dent = { };
+	int n = 0;
+	int first_free = -1;
+
+	for (n = 0;; n++) {
+		res = read_dent(dirh, n, &dent);
+		if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+			memset(&dent, 0, sizeof(dent));
+			if (first_free != -1)
+				n = first_free;
+			break;
+		}
+		if (res)
+			return res;
+
+		if (is_free(&dent) && first_free == -1)
+			first_free = n;
+	}
+
+	dfh->idx = n;
+	dfh->file_number = dent.file_number;
```

Comment:
"file_number" and "hash" is always ignored by the caller and always invalid since they are from a free entry.